### DISCUSSION TITLE
Bootstrap Update

### DISCUSSION THREAD

**Discussion by danemorgan** [holger1411/understrap#1275] [General]
Does anyone have any comments, ideas, warnings, concerns, advice regarding updating to Bootstrap 5?

**Comment by bacoords** [comment]
Can we still support BS4 so as not to break anything? If so, some thoughts:
1. We could try to keep both 4 & 5 as options, perhaps as a setting in the customizer (Beaver Builder Theme did something similar). Then we can have a filter that returns the current BS version and could potentially be overridden in a child theme.
2. A filter (similar to the old HTML5 transition days) could be used to determine the correct usage of any classes, data attributes, navwalker, or other pieces that have would have to be different.
3. Can we make BS5 the default for new installs, but BS4 the default for currently running themes so we don't break any customizations? Last I recall, themes don't have a similar 'activation' hook as plugins.

**Reply by pattonwebz** [reply]
The navwalker in the theme doesn't support BS5. There are some issue threads on the navwalker repo that say how it could be made to support it but they seem a bit hackish.

The walker probably needs a big refactor/refinement to properly support BS5 and bring it in line with changes to core navwaler. I have barely touched the code beyond minor fixes in probably 4ish years but I'm open to accepting pull requests on the repo of they are tested :)

**Reply by pattonwebz** [reply]
The `after_switch_theme` hook can be used in a theme in a similar fashion as plugin activation hooks.

**Reply by 0dp** [reply]
> The navwalker in the theme doesn't support BS5. There are some issue threads on the navwalker repo that say how it could be made to support it but they seem a bit hackish.
> 
> The walker probably needs a big refactor/refinement to properly support BS5 and bring it in line with changes to core navwaler. I have barely touched the code beyond minor fixes in probably 4ish years but I'm open to accepting pull requests on the repo of they are tested :)

By coincidence i just stumbled over this project:
https://github.com/AlexWebLab/bootstrap-5-wordpress-navbar-walker/
Maybe it could provide some insight 🕵️‍♂️

**Comment by danemorgan** [comment]
I notice that we have a [Stable Bootstrap3 version of UnderStrap](https://github.com/understrap/understrap-bs3). We were talking about creating a configuration or switch or detecting previous installs through a theme activation check, but I wonder if it might be better to create a Stable Bootstrap 4 version in the same way and just move forward with Bootstrap5. There is something clean in that approach that I like.

**Reply by 0dp** [reply]
I like it

**Comment by Thomas-A-Reinert** [comment]


**Reply by danemorgan** [reply]
Do you have major updates on auto-update? We are talking about a major update here, not a minor. Going forward, I want to use semantic versioning for that reason. There *will* be breaking updates. While there *are* non-developers using the theme and while we should work to help them, they are not the demographic for this theme.

How long do we keep switches? Do we still try to have BS4 built-in when we move to BS6? BS7? Is there a line anywhere?

**Reply by Thomas-A-Reinert** [reply]
Is there any option distinguising minor from major updates? Afaik: Nope.
(Could be done with an NPM-Script though)

As I said in the first place - get a roadmap up and runnig. With something that can be LINKED to in any case.
Nothing happened. And who you wanna support - Devs, Usersl People who just like that blank theme?. Nope. Nothing yer. 

So the BS1-2-3-4-5-6-7-7-8-X Support is up to YOU not ME or anyone else!

But anyone using the actual version should KNOW what´s happening WHEN and WHERE.

**Reply by danemorgan** [reply]
Our current roadmap Is basically this thread. Our current roadmap is updating to BS5. You're in it discussing it now.

**Reply by danemorgan** [reply]
Looking into it, no the themes and plugins do not distinguish between minor and major updates. I've never used automatic updates for themes or plugins. I had just assumed they followed core with minor updates opt-out and major updates opt-in.

I'm still not sure I'm following you here though.

I don't think it likely that people who are making extensive edits to a child theme are downloading the theme from the repository and turning on updates. They are going to see a new release issued if they even subscribe and read the changelog before they just pull the main.

People who are extensively editing the theme itself aren't going to update at all.

People who are installing from the repository and just using it will get the theme code updated for the Bootstrap changes.

People who are installing from the repository and making minor edits to a child theme may update ( if they turned on auto-updates or noticed there was an update available ) and may experience breakage.

Yes, we want to mitigate that to what extent is reasonable, but that isn't the only factor to consider.

Maybe you aren't talking about auto-updates from the theme repository, but something else?

**Reply by bacoords** [reply]
I tend to agree that there's going to be a non-negligible group of users who are using the parent theme from the dotOrg repo and a child theme that assumes the underlying html structures are still based on BS4. I've done this on plenty of sites for clients. These clients may or may not be autoupdating but they may not understand the meaning of updating to BS5 and that it might break their custom stylesheets. 

To swap out the parent theme to BS5 without making it opt-in or at least having a path to staying on BS4 (with ample dashboard admin notices explaining it) doesn't seem right and will probably lead to some broken sites and backlash. 

I see a few paths:

- Continue supporting BS4 but make BS5 opt-in, possibly also default for new installs, as say a customizer setting (Beaver Builder theme did something similar from BS3 to 4. The main difference is that BS5 includes a lot more changes to structural markup. It's a much bigger conceptual jump). 
- Fork this repo to a BS4 "legacy" version that continues to feed minor updates to the dotOrg community. Use this repo to migrate up to BS5 and submit it as a brand new parent theme (i.e. Understrap 5) to dotOrg. Maybe include an admin notice in the original theme that a new BS5 version of the theme is available IF they want to test it out. 
- Update this theme to BS5 and tell all child theme users they may have to stay on the 0.9.x branch if they want their site to work. Maybe release a version with an admin notice that links to a deep explainer on it. 

There's major pros and major cons to the various options, which is why I'm grateful for all contributions to this discussion.

**Reply by danemorgan** [reply]
@bacoords 

> * Fork this repo to a BS4 "legacy" version that continues to feed minor updates to the dotOrg community. Use this repo to migrate up to BS5 and submit it as a brand new parent theme (i.e. Understrap 5) to dotOrg. Maybe include an admin notice in the original theme that a new BS5 version of the theme is available IF they want to test it out.

This is where I'm leaning.

If we create a legacy BS4 repository, we can continue with any bug fixes or security concerns that come along, but consider it feature complete and no longer under development, and the main Understrap repository can move on cleanly is what I had already come up with. 

Going through the documentation on what's different, trying to support both BS4 and BS 5 in a single theme would practically be two themes with an internal switch to pick between them. The changes will be extensive throughout.

Submitting a new theme to the repository for the update was the part I was missing. That keeps each clean and simple and feels, to me, in keeping with the WordPress philosophy document.

@pattonwebz, I know there is some guidance about similar themes in the repository. Does a fork like this work to get through the theme inspection process?

**Comment by Thomas-A-Reinert** [comment]


**Reply by danemorgan** [reply]
Closed channels don't make sense to me. They limit transparency.

How we handle BS4 as we move to BS5, and how our method for doing so affects different users *is* the topic. Your input is desired, but so is everyone else's.

Nothing has been decided here, much less dictated. A discussion has been opened.

I think we can have an open discussion about it without yelling and belligerence. for my part, that is how I intend to proceed.

**Comment by 0dp** [comment]
I merely want to propose that the BS5 roll out does not commence before we have a new and improved build system in place and that we clean up the theme directory and drop all dev stuff into /src

**Reply by danemorgan** [reply]
I concur. Build system first, then Bootstrap.

**Reply by ironicmoka** [reply]
As stated [on a issue in the child theme repo](https://github.com/understrap/understrap-child/issues/239) I think that having a theme or child that can be installed and working up and running without having to provision or start local gulp tasks is crucial to get on board the less skilled user base. What I loved about understrap was the very low entry point for newbies like me. Even today I maintain custom client projects created with understrap child where I never have run gulp one single time: just compiled the css with a VSCode extension on updates and relied to the src folder for what already was there.

Apart from my personal experience I think that having all the dependencies already packed up could prevent confusion for new and old users and less friction for new adopters. I know that a framework should primarly be stable, clean and reliable but the WordPress community is especially known for its accessibility for everyone: I started using Understrap as my first starter theme because of its skill scalability and because when I installed it the very first time I knew nothing about npm or gulp but that didn't stopped me to make it work the way I needed.

I hope I'm not missing the point here with this feedback: maybe I overestimate the consequences of getting rid of the src folder and underestimate the benefit but I think my point of view could be useful to plan ahead minding all the not yet skilled wannabe developer that can learn the ropes with a tool easy enough to set up working out of the box but flexible for more advanced devs.

**Reply by danemorgan** [reply]
Incidental to this is an idea I have in the back of my mind that I would like to try to get the child theme into the dot org repository at some point.

In order to do that all of the build process has to be removed. The src would be critical for users wanting a starting point to edit a child theme from in that case.

**Reply by 0dp** [reply]
To clarify @ironicmoka  and any doubts you might have.

The theme will always work without the use of build tools and developer forward technologies.
It's a requirement by wp.org anyways; no frameworks allowed.

So you'd always be able to develop the theme without using the build tools and just using common wp approaches like enqueued custom styles and scripts, the customizer, widgets, plugins and so on.

**Comment by 0dp** [comment]
This is epic https://blog.getbootstrap.com/2021/08/04/bootstrap-5-1-0/

**Reply by danemorgan** [reply]
I was very happy last night when I saw that CSS Grid is coming in a soonly kind of a way.

**Reply by 0dp** [reply]
Exactly
